docs/architecture/adr-036-blockchain-reactor-refactor.md

```diff
@@ -0,0 +1,406 @@
+
+					if _, ok := state.PeerMap[event.PeerID]; ok {
+						msg = createBlockRequestMessage(state, event.PeerID, peerStats.Height)
+						if msg != nil {
+							peerStats.PendingRequests = msg.Height
+							state.PendingRequestsNum++
+							timeout = ResponseTimeoutTrigger{msg.PeerID, msg.Height, PeerTimeout}
+						} else if state.PendingRequestsNum == 0 {
+							timeout = TerminationTimeoutTrigger{state.Height, TerminationTimeout}
+						}
+					}
+				} else {
+					error = "Received Block from wrong peer!"
+				}
+			} else {
+				error = "Received Block from unknown peer!"
+			}
+
+			state.PeerMap[event.PeerID] = peerStats
+			return state, msg, timeout, error
+
+		case EventResponseTimeout:
+			if _, ok := state.PeerMap[event.PeerID]; ok {
+				peerStats = state.PeerMap[event.PeerID]
+				if peerStats.PendingRequest == event.Height {
+					add(state.FailedRequests, pendingRequest)
+					delete(state.PeerMap, event.PeerID)
+					state.PendingRequestsNum--
+					if state.PeerMap.isEmpty() || state.PendingRequestsNum == 0 {
+						timeout = TimeoutTrigger{state.Height, TerminationTimeout}
+					}
+				}
+			}
+
+			return state, msg, timeout, error
+
+		case EventTerminationTimeout:
+			if state.Height == event.Height && state.PendingRequestsNum == 0 {
+				state.State = ConsensusMode
+			}
+			return state, msg, timeout, error
+
+		default:
+			error = "Received unknown event type!"
+			return state, msg, timeout, error
+		}
+
+	}
+}
+
+func createBlockResponseMessage(state ControllerState, event BlockRequest) BlockResponseMessage {
+	msg = nil
+	if _, ok := state.PeerMap[event.PeerID]; !ok {
+		peerStats = PeerStats{-1, -1}
+	}
+	if state.Executor.ContainsBlockWithHeight(event.Height) && event.Height > peerStats.Height {
+		peerStats = event.Height
+		msg = BlockResponseMessage{
+			Height:        event.Height,
+			Block:         state.Executor.getBlock(eventHeight),
+			Commit:        state.Executor.getCommit(eventHeight),
+			PeerID:        event.PeerID,
+			CurrentHeight: state.Height - 1,
+		}
+	}
+	state.PeerMap[event.PeerID] = peerStats
+	return msg
+}
+
+func createBlockRequestMessage(state ControllerState, peerID ID, peerHeight int64) BlockRequestMessage {
+	msg = nil
+	blockNumber = -1
+
+	// exist request r in state.FailedRequests such that r <= peerHeight
+	if true {
+		blockNumber = r
+		delete(state.FailedRequests, r)
+	} else if state.MaxRequestPending < peerHeight {
+		state.MaxRequestPending++
+		blockNumber = state.MaxRequestPending // increment state.MaxRequestPending and then return the new value
+	}
+
+	if blockNumber > -1 {
+		msg = BlockRequestMessage{blockNumber, peerID}
+	}
+	return msg
+}
+
+func verifyBlocks(state State) State {
+	done = false
+	for !done {
+		block = state.Store[height]
+
+		if block != nil {
+			// TODO: verify block.Block using block.Commit
+			verified = true
+
+			if verified {
+				block.Execute()
```

Comment:
step of `block.Execute()` takes most of the time of state transition. Since state transition is serial. We may find a way to let `block.Execute()` tasks run async.

Reply:
Agree. Verifying and executing blocks are good candidates to be isolated in a separate routine but as stated above the idea is to focus on performance improvements after (if needed).